src/clj/puppetlabs/puppetserver/jruby_request.clj

```diff
@@ -45,6 +46,24 @@
      jruby-puppet jruby-service {:request (dissoc request :ssl-client-cert)}
      (handler (assoc request :jruby-instance jruby-puppet)))))
 
+(defn wrap-with-concurrent-request-limit
+  "Middleware fn that returns a 503 'Service Unavailable' response if the
+  number of outstanding requests for a JRuby instance exceed the given limit.
+  If splay-limit is set, the response will include a Retry-After header set
+  to a random fraction of the limit."
+  [handler metrics-service request-limit splay-limit]
+  ;; TODO: Log warnings when limit is hit. Add i18n to user-facing strings.
+  (fn [request]
+    (let [metrics-map (jruby-metrics/get-metrics metrics-service)
+          requested-count (-> metrics-map :requested-count .getCount)]
+      (if (> requested-count request-limit)
+        (let [response (ringutils/plain-response 503 "JRuby capacity exceeded.")]
+          (if (> splay-limit 0)
+            (assoc-in response [:headers "Retry-After"]
```

Comment:
This is an interesting idea - controlling the splay from the server side.  Does the Puppet agent not already have any concept of a splay / randomization built into it to retry a request based on the receipt of specific HTTP error codes, like a 503?  This could get more interesting in the future if we could tune the server to more heuristically determine a good retry threshold to use based on load, rather than just always using a random value.  I think what you have would be a good approach for now, though.